src/VecSim/algorithms/brute_force/bf_batch_iterator.h

```diff
@@ -98,25 +98,23 @@ VecSimQueryResult_List BF_BatchIterator<DataType, DistType>::heapBasedSearch(siz
     DistType upperBound = std::numeric_limits<DistType>::lowest();
     vecsim_stl::max_priority_queue<DistType, labelType> TopCandidates(this->allocator);
     // map vector's label to its index in the scores vector.
-    vecsim_stl::unordered_map<size_t, size_t> TopCandidatesIndices(n_res, this->allocator);
+    vecsim_stl::unordered_map<labelType, size_t> TopCandidatesIndices(n_res, this->allocator);
     for (size_t i = this->scores_valid_start_pos; i < this->scores.size(); i++) {
-        if (TopCandidates.size() < n_res) {
-            TopCandidates.emplace(this->scores[i].first, this->scores[i].second);
-            TopCandidatesIndices[this->scores[i].second] = i;
-            upperBound = TopCandidates.top().first;
-        } else {
-            if (this->scores[i].first >= upperBound) {
-                continue;
-            } else {
+        if (TopCandidates.size() >= n_res) {
+            if (this->scores[i].first < upperBound) {
                 // remove the furthest vector from the candidates and from the label->index mappings
                 // we first remove the worst candidate so we wont exceed the allocated size
                 TopCandidatesIndices.erase(TopCandidates.top().second);
                 TopCandidates.pop();
-                TopCandidatesIndices[this->scores[i].second] = i;
-                TopCandidates.emplace(this->scores[i].first, this->scores[i].second);
-                upperBound = TopCandidates.top().first;
+            } else {
+                continue;
             }
         }
+        // If top candidate heap size is smaller than required results number
+        // or the current score is closer than the farthermost vector's -> insert current vector
```

Comment:
This comment phrasing is misleading. If we got to here, the heap size is < n since it is either in filling up or was poped because we found a better vector.